### PR TITLE
Cache will replicate

### PR DESCRIPTION
**Pull request by pinzart90**
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

(FILL ME IN) This section describes why this PR is here. Usually it would include a reference 
to the tracking task that it is part or all of the solution for.

### Declarations

Check these if you believe they are true

- [ ] The codebase is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [ ] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Release Notes

(FILL ME IN) Brief description of the fix / enhancement. **Mandatory section** 

### Reviewers

(FILL ME IN) Reviewer 1  (If possible, assign the Reviewer for the PR)

(FILL ME IN, optional) Any additional notes to reviewers or testers.

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

### CODE REVIEW

File: src/Engine/ProtoScript/Runners/LiveRunner.cs [hunk 1]
```diff
@@ -1270,13 +1270,11 @@ public void Dispose()
...
             Dictionary<string, IList> input = new Dictionary<string, IList>();
-            var assemblyPath = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location));
 
+            var assemblyPath = Path.Combine(Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location));
             //TODO_MSIL: remove the dependency on the old VM by implementing
             //necesary Emit functions(ex mitFunctionDefinition and EmitImportStatements and all the preloading logic)
-            msilRuntimeCore = msilRuntimeCore ?? new MSILRuntimeCore(runtimeCore);
-            var codeGenIL = new EmitMSIL.CodeGenIL(input, Path.Combine(assemblyPath, "opCodes.txt"), msilRuntimeCore);
-
+            codeGenIL = codeGenIL ?? new EmitMSIL.CodeGenIL(input, Path.Combine(assemblyPath, "opCodes.txt"), new MSILRuntimeCore(runtimeCore));
```
Comment (pinzart90) [inline src/Engine/ProtoScript/Runners/LiveRunner.cs:1277]: re-using codeGen caches for every execution

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 2]
```diff
@@ -101,41 +105,43 @@ internal CodeGenIL(IDictionary<string, IList> input, string filePath, ProtoCore.
...
-            var asm = BuilderHelper.CreateAssemblyBuilder("DynamicAssembly", false, access);
-            // 2. Create module builder
-            var mod = BuilderHelper.CreateDLLModuleBuilder(asm, "DynamicAssembly");
-            // 3. Create type builder (name it "ExecuteIL")
-            var type = BuilderHelper.CreateType(mod, "ExecuteIL");
-            // 4. Create method ("Execute"), get ILGenerator 
-            var execMethod = BuilderHelper.CreateMethod(type, "Execute",
-                System.Reflection.MethodAttributes.Static | System.Reflection.MethodAttributes.Private, typeof(void), new[] { typeof(IDictionary<string, IList>),
+            writer = new StreamWriter(logPath);
+            using (Disposable.Create(() => { writer.Close(); writer.Dispose(); }))
```
Comment (mjkkirschner) [inline src/Engine/EmitMSIL/CodeGenIL.cs:109]: from what I can find online, dispose will also close the stream, so I think you're safe to just use a regular using statement over the writer.

File: src/Engine/ProtoCore/Lang/FunctionEndPoint.cs [hunk 3]
```diff
@@ -46,6 +46,10 @@ internal List<ParamInfo> FormalParams
 
         public MemberInfo MemberInfo => method.Info;
 
+        private string sig = string.Empty;
+
+        internal string Signature => sig;
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/FunctionEndPoint.cs:51]: why do we need the field?

File: src/Engine/ProtoCore/Lang/FunctionEndPoint.cs [hunk 4]
```diff
@@ -66,6 +70,25 @@ internal CLRFunctionEndPoint(FFIMemberInfo method, List<ParamInfo> formalParams,
             this.method = method;
             this.FormalParams = formalParams;
             this.procedureNode = procedureNode;
+
+            if (method.Info is MethodInfo mInfo)
+            {
+                sig += mInfo.ReturnType.FullName + " " + mInfo.DeclaringType.FullName + ":" + mInfo.Name;
```
Comment (mjkkirschner) [inline src/Engine/ProtoCore/Lang/FunctionEndPoint.cs:76]: IMO $"{}" interpolation is easier to read.

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 5]
```diff
@@ -1691,25 +1697,33 @@ private static int GetArgumentRank(AssociativeNode val)
...
         }
 
-        private bool WillCallReplicate(List<Type> paramTypes, bool isStaticOrCtor, List<AssociativeNode> args)
+        private bool WillCallReplicate(string fepSig, List<Type> paramTypes, bool isStaticOrCtor, List<AssociativeNode> args)
         {
-            if (isStaticOrCtor)
+            if (willReplicateCache.TryGetValue(fepSig, out bool willReplicate))
             {
-                return !DoesParamArgRankMatch(paramTypes, args);
+                return willReplicate;
```
Comment (mjkkirschner) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1704]: so - if the arguments change won't this value potentially be incorrect?

I feel that either
1. We also need to put the arguments into the signature as the key, OR
2. We need to fallback to replication when making a direct call, maybe using a try catch that gets emitted?

**Review by aparajit-pratap — CHANGES_REQUESTED** [review]
I'm not sure I understand this. The decision to replicate or not (true or false) depends on not just the function signature but also the arguments used in calling that function. It could be that in one case, the call replicates, while in the other it doesn't for the same signature, so I don't believe that using the function signature as the key is going to serve the purpose.

**Comment by aparajit-pratap** [comment]
I tried using the AST node ID as the key for this cache (see commented code [here](https://github.com/DynamoDS/Dynamo/blob/f787a2481f27cf40d69faab27c0cdc8db42ae39f/src/Engine/EmitMSIL/CodeGenIL.cs#L1336)), the only problem with this approach is that the node ID changes midway as some ASTs are recreated between compilation passes, which isn't expected. I believe one way to do that is to also cache those ASTs so that they can be reused between compilation passes instead of recreating them again.

File: src/Engine/EmitMSIL/CodeGenIL.cs (hunk 5 above)
Comment (pinzart90) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1704]: Yep, totally forgot about the arguments

**Comment by aparajit-pratap** [comment]
> InheritID

@mjkkirschner to use this, won't we need to cache the previously created ASTs to inherit their IDs anyway, in which case, we might as well, just skip the creation step and reuse these ASTs?

**Comment by pinzart90** [comment]
@aparajit-pratap the only place I see that creates new FunctionCallNodes on the fly is in `EmitRangeExpression`. Weca stabilize the generated funcCall's ID by inheriting from the RangeExpressionNode. Do you think that will work?

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 6]
```diff
@@ -1883,6 +1879,12 @@ private Type EmitRangeExprNode(AssociativeNode node)
 
 
             var rangeExprFunc = AstFactory.BuildFunctionCall(methodName, arguments);
+
+            // We need to keep the ID of the generated fuction call stable
+            // in order to make use of caches in between different compile stages.
+            // So just copy it from the rangeExpression node
+            rangeExprFunc.InheritID(range.ID);
```
Comment (pinzart90) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1886]: this is the only `InheritID` used.

Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1885]: So this is AST is still created twice - once in `gatherTypeInfo` and once in `emitIL` phase?

Comment (pinzart90) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1885]: yes

Comment (aparajit-pratap) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1885]: I'm okay with this if you think that caching the AST will be too much work for little benefit.

File: src/Engine/EmitMSIL/CodeGenIL.cs [hunk 7]
```diff
@@ -1337,26 +1338,8 @@ private Type EmitFunctionCallNode(AssociativeNode node)
...
-            //        throw new Exception($"ast {node.ID}:{node.Kind} already exists in replicated call map.");
-            //    }
-            //    isRepCall.Add(node.ID, WillCallReplicate(parameters, isStatic, args));
-            //}
-
-            //bool doesReplicate = false;
-            //if(!isRepCall.TryGetValue(node.ID, out doesReplicate))
-            //{
-            //    throw new Exception($"ast { node.ID }:{ node.Kind} does not exist in replicated call map.");
-            //}
```
Comment (aparajit-pratap) [inline src/Engine/EmitMSIL/CodeGenIL.cs:1353]: Please delete commented line 27 too.

**Review by aparajit-pratap — APPROVED** [review]
one question and one comment, then LGTM

File: src/Engine/EmitMSIL/CodeGenIL.cs (hunk 6 above)
Comment (pinzart90) [reply src/Engine/EmitMSIL/CodeGenIL.cs:1885]: the cost to recreate the AST for `rangeExprFunc` seems to be very low

**Pull request merged**